lib/src/block_parser.dart

```diff
@@ -107,11 +111,25 @@ class BlockParser {
 
   List<Node> parseLines() {
     final blocks = <Node>[];
+
+    // If the `_pos` does not change before and after `parse()`, never try to
+    // parse this `_pos` with the same syntax again.
+    BlockSyntax? neverMatch;
```

Comment:
Apologies, I've read through this a few times and I can't get my head around this variable. Specifically it's name, and what the check on line 121 is for. Can you add a comment maybe on line 121 and 131. I just know that when I look at this code months from now, I'll have no idea what [neverMatch] is for.

Reply:
Because of this `retreat`, which moves the pointer one step back, the result will be an infinite loop if without this `neverMatch` mechanism. 

https://github.com/dart-lang/markdown/blob/03e82eec45abd165f3959a95ef0a8722cc2bafb2/lib/src/block_syntaxes/table_syntax.dart#L39

Reply:
Normally in `canParse` we should make sure a structure satisfies all the requirements of the target element before ending up into `parse` method. But for some elements such as tables,  their tokens locate at the second line, they might  also relay on some other conditions or even the type of the following element to make them a valid target element. It might be low performance if we put all these validations in `canParse` method,  except we cache some calculated result from `canParse` and reuse them in `parse` method. 

With the `retreat` and `neverMatch` mechanism,  `canParse` can do a roughly match,  the `parse` has a chance to regret the `true` result from `canParse`.

I might implement the same mechanism in setext headings and link reference definitions later if it has a better performance.

Reply:
Thanks. I also added one more comment to this line.